Refactor jitMemcpy into atomic and non-atomic versions

justinmichaud opened the pull request:
#### 952e58054d0fd2b45264d09bf2ef391d1ddb1f3b
<pre>
Refactor jitMemcpy into atomic and non-atomic versions
<a href="https://bugs.webkit.org/show_bug.cgi?id=293649">https://bugs.webkit.org/show_bug.cgi?id=293649</a>

Reviewed by Keith Miller.

On ARMv7, there are some concurrent repatching issues because of
unaligned writes to the jit region. Instead of repatching atomically whenever it
is easy, we should be explicit about when we expect writes to the jit
region to tear or not.

Canonical link: <a href="https://commits.webkit.org/299398@main">https://commits.webkit.org/299398@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/799925395153ebde804254105216fc6ee91f4890

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118661 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38342 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124841 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70721 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46924 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90064 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59596 "Found 1 new test failure: http/tests/navigation/ping-attribute/anchor-cookie.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31103 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106393 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24504 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68498 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110779 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100542 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24694 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127899 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117175 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45568 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34389 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98701 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45932 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102613 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98482 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25076 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43925 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21929 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42073 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45438 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51116 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145871 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44902 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37505 "Found 19 new JSC stress test failures: wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.default-wasm, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-bbq, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-bbq-no-consts, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-collect-continuously, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-eager, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-eager-jettison, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-no-cjit, wasm.yaml/wasm/js-api/memory-toResizableBuffer.js.wasm-slow-memory, wasm.yaml/wasm/lowExecutableMemory/executable-memory-oom.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48248 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46588 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -44,6 +44,7 @@ using RepatchingInfo = WTF::ConstexprOptionSet<RepatchingFlag>;
 static constexpr RepatchingInfo jitMemcpyRepatch = RepatchingInfo { };
 static constexpr RepatchingInfo jitMemcpyRepatchAtomic = RepatchingInfo { RepatchingFlag::Atomic };
 static constexpr RepatchingInfo jitMemcpyRepatchFlush = RepatchingInfo { RepatchingFlag::Flush };
+static constexpr RepatchingInfo jitMemcpyAtomicRepatchFlush = RepatchingInfo { RepatchingFlag::Atomic, RepatchingFlag::Flush };
```
nit: can we call this `jitMemcpyRepatchAtomicAndFlush` or `jitMemcpyRepatchAtomicFlush` to match the other values above which all start with `jitMemcpyRepatch`?

Can you also add a comment above the definition of `RepatchingFlag` above to explain what the intent of this flag is, as in when should we use `Atomic`, `Memcpy`, and `Flush` ... or what do they actually mean in the context they are supposed to be used?  Thanks.

MenloDorian commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -44,6 +44,7 @@ using RepatchingInfo = WTF::ConstexprOptionSet<RepatchingFlag>;
 static constexpr RepatchingInfo jitMemcpyRepatchFlush = RepatchingInfo { RepatchingFlag::Flush };
+static constexpr RepatchingInfo jitMemcpyAtomicRepatchFlush = RepatchingInfo { RepatchingFlag::Atomic, RepatchingFlag::Flush };
 
 constexpr RepatchingInfo noFlush(RepatchingInfo i)
```
Not due to this patch but the operation to filter out `jitMemcpyRepatchFlush` here is using a minus operation instead of a mask. This can be problematic if `noFlush` is ever used on a value that doesn't already have jitMemcpyRepatchFlush set.  Since this is a constexpr function anyway, let's make it robust by using a proper mask instead of the minus operation.

justinmichaud replied:
Minus is overloaded, that's why I used it. I can do a mask instead

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3587,13 +3587,13 @@ class ARM64Assembler {
     {
         ASSERT(from.isSet());
-        relinkJumpOrCall<BranchType_JMP>(addressOf(code, from), addressOf(code, from), to);
+        relinkJumpOrCall<BranchType_JMP, jitMemcpyRepatch>(addressOf(code, from), addressOf(code, from), to);
```
Why does `ARM64Assembler.h` need this change?  The tearing issue only applies to ARMv7, correct?  relinkJumpOrCall is both defined and used only in `ARM64Assembler.h`.  This is not needed.

kmiller68 replied:
@justinmichaud Can you record the comment you made in slack here for posterity?

justinmichaud replied:
Done! I put it in relinkJumpOrCall

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3587,13 +3587,13 @@ class ARM64Assembler {
     {
         ASSERT(from.isSet());
-        linkJumpOrCall<BranchType_CALL>(addressOf(code, from) - 1, addressOf(code, from) - 1, to);
+        linkJumpOrCall<BranchType_CALL, jitMemcpyRepatch>(addressOf(code, from) - 1, addressOf(code, from) - 1, to);
```
Ditto.  The same goes for `linkJumpOrCall`.  This change is not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3629,9 +3629,11 @@ class ARM64Assembler {
     {
-        fillNops<jitMemcpyRepatch>(where, memoryToFillWithNopsInBytes);
+        static_assert((*repatch).contains(RepatchingFlag::Flush));
+        fillNops<noFlush(repatch)>(where, memoryToFillWithNopsInBytes);
```
This change is not needed, right?  We can just ignore the incoming `RepatchingInfo`.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3712,21 +3714,26 @@ class ARM64Assembler {
     {
-        relinkJumpOrCall<BranchType_JMP>(reinterpret_cast<int*>(from), reinterpret_cast<const int*>(from), to);
+        static_assert((*repatch).contains(RepatchingFlag::Flush));
+        relinkJumpOrCall<BranchType_JMP, noFlush(repatch)>(reinterpret_cast<int*>(from), reinterpret_cast<const int*>(from), to);
```
Ditto.  Not needed.  Just ignore the incoming `RepatchingInfo`.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3712,21 +3714,26 @@ class ARM64Assembler {
     {
-        relinkJumpOrCall<BranchType_CALL>(reinterpret_cast<int*>(from) - 1, reinterpret_cast<const int*>(from) - 1, to);
+        static_assert((*repatch).contains(RepatchingFlag::Flush));
+        relinkJumpOrCall<BranchType_CALL, noFlush(repatch)>(reinterpret_cast<int*>(from) - 1, reinterpret_cast<const int*>(from) - 1, to);
```
Ditto.  Not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3961,7 +3968,7 @@ class ARM64Assembler {
         setPointer<repatch>(address, valuePtr, rd);
     }
 
-    template<BranchType type, RepatchingInfo repatch = jitMemcpyRepatch>
+    template<BranchType type, RepatchingInfo repatch>
     static void linkJumpOrCall(int* from, const int* fromInstruction, void* to)
```
`RepatchingInfo` is not needed here at all and can be removed, right?  This `linkJumpOrCall` is purely used inside ARM64Assembler.h itself.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3996,7 +4003,7 @@ class ARM64Assembler {
         machineCodeCopy<repatch>(from, &insn, sizeof(int));
     }
 
-    template<BranchTargetType type, RepatchingInfo repatch = jitMemcpyRepatch>
+    template<BranchTargetType type, RepatchingInfo repatch>
     static void linkCompareAndBranch(Condition condition, bool is64Bit, RegisterID rt, int* from, const int* fromInstruction, void* to)
```
Ditto.  `RepatchingInfo` is not needed and can be removed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4022,7 +4029,7 @@ class ARM64Assembler {
         }
     }
 
-    template<BranchTargetType type, RepatchingInfo repatch = jitMemcpyRepatch>
+    template<BranchTargetType type, RepatchingInfo repatch>
     static void linkConditionalBranch(Condition condition, int* from, const int* fromInstruction, void* to)
```
Ditto.  `RepatchingInfo` is not needed and can be removed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4048,7 +4055,7 @@ class ARM64Assembler {
         }
     }
 
-    template<BranchTargetType type, RepatchingInfo repatch = jitMemcpyRepatch>
+    template<BranchTargetType type, RepatchingInfo repatch>
     static void linkTestAndBranch(Condition condition, unsigned bitNumber, RegisterID rt, int* from, const int* fromInstruction, void* to)
```
Ditto.  `RepatchingInfo` is not needed and can be removed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4075,7 +4082,7 @@ class ARM64Assembler {
 
-    template<BranchType type>
+    template<BranchType type, RepatchingInfo repatch>
     static void relinkJumpOrCall(int* from, const int* fromInstruction, void* to)
```
Ditto.  `RepatchingInfo` is not needed and can be removed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4092,7 +4099,7 @@ class ARM64Assembler {
                 if (imm19 == 8)
                     condition = invert(condition);
 
-                linkConditionalBranch<IndirectBranch>(condition, from - 1, fromInstruction - 1, to);
+                linkConditionalBranch<IndirectBranch, repatch>(condition, from - 1, fromInstruction - 1, to);
```
Ditto not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4105,7 +4112,7 @@ class ARM64Assembler {
                 if (imm19 == 8)
                     op = !op;
 
-                linkCompareAndBranch<IndirectBranch>(op ? ConditionNE : ConditionEQ, opSize == Datasize_64, rt, from - 1, fromInstruction - 1, to);
+                linkCompareAndBranch<IndirectBranch, repatch>(op ? ConditionNE : ConditionEQ, opSize == Datasize_64, rt, from - 1, fromInstruction - 1, to);
```
Ditto.  Not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4117,12 +4124,12 @@ class ARM64Assembler {
                 if (imm14 == 8)
                     op = !op;
 
-                linkTestAndBranch<IndirectBranch>(op ? ConditionNE : ConditionEQ, bitNumber, rt, from - 1, fromInstruction - 1, to);
+                linkTestAndBranch<IndirectBranch, repatch>(op ? ConditionNE : ConditionEQ, bitNumber, rt, from - 1, fromInstruction - 1, to);
```
Ditto.  Not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -4117,12 +4124,12 @@ class ARM64Assembler {
                 if (imm14 == 8)
                     op = !op;
 
-                linkTestAndBranch<IndirectBranch>(op ? ConditionNE : ConditionEQ, bitNumber, rt, from - 1, fromInstruction - 1, to);
+                linkTestAndBranch<IndirectBranch, repatch>(op ? ConditionNE : ConditionEQ, bitNumber, rt, from - 1, fromInstruction - 1, to);
                 return;
             }
         }
 
-        linkJumpOrCall<type>(from, fromInstruction, to);
+        linkJumpOrCall<type, repatch>(from, fromInstruction, to);
```
Ditto.  Not needed.

MenloDorian commented on Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
```diff
@@ -1014,21 +1014,23 @@ class AbstractMacroAssembler : public AbstractMacroAssemblerBase {
     {
-        AssemblerType::relinkJump(jump.dataLocation(), destination.dataLocation());
+        constexpr auto repatch = atomic ? jitMemcpyAtomicRepatchFlush : jitMemcpyRepatchFlush;
+        AssemblerType::template relinkJump<repatch>(jump.dataLocation(), destination.dataLocation());
```
It seems that the real need is only to convey whether this needs to be atomic or not.  Is there really a need to use the `RepatchingInfo` abstraction?  I wonder if it is something that should be constrained to the ARMv7 only implementation.  It's fair to have the high level shared code here pass in that atomicity is required, but it seems that flushing is implied to always be needed here.  I wonder if this is also true elsewhere.

Note: if atomicity is really the only unique piece of info needed, then it makes it clear why ARM64 and X86_64 ports can completely ignore the `atomic` template parameter, because their repatching are already always atomic.  It'd be nice if that's how it plays out.

MenloDorian commented on Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
```diff
@@ -1014,21 +1014,23 @@ class AbstractMacroAssembler : public AbstractMacroAssemblerBase {
+    template<bool atomic = false, PtrTag callTag, PtrTag destTag>
     static void repatchNearCall(CodeLocationNearCall<callTag> nearCall, CodeLocationLabel<destTag> destination)
     {
+        constexpr auto repatch = atomic ? jitMemcpyAtomicRepatchFlush : jitMemcpyRepatchFlush;
```
Ditto here.  It seems the only real piece of info needed is the atomicity.

MenloDorian commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -5211,7 +5211,7 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
     static void reemitInitialMoveWithPatch(void* address, void* value)
     {
-        Assembler::setPointer<jitMemcpyRepatchFlush>(static_cast<int*>(address), value, dataTempRegister);
+        Assembler::setPointer<jitMemcpyAtomicRepatchFlush>(static_cast<int*>(address), value, dataTempRegister);
```
`setPointer` is not used by common code.  So, ARM64 an X86_64 port shouldn't need it to be parameterized.  Perhaps the parameter should be removed to simplify the code.

MenloDorian commented on Source/JavaScriptCore/assembler/ARMv7Assembler.h:
```diff
@@ -2775,23 +2784,25 @@ class ARMv7Assembler {
     {
-        fillNops<jitMemcpyRepatch>(instructionStart, memoryToFillWithNopsInBytes);
+        static_assert((*repatch).contains(RepatchingFlag::Flush));
+        fillNops<noFlush(repatch)>(instructionStart, memoryToFillWithNopsInBytes);
```
Same detail here.  It seems the only unique piece of info we need to convey is the atomicity.  All implementations of `replaceWithNops` already require cache flushing.  So, it was superfluous to have to pass it in, and then to have to `static_assert` that it was passed in.  Is `RepatchingInfo` really needed, or can we just do with an atomicity requirement?

kmiller68 commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3712,21 +3714,26 @@ class ARM64Assembler {
-        relinkJumpOrCall<BranchType_JMP>(reinterpret_cast<int*>(from), reinterpret_cast<const int*>(from), to);
-        cacheFlush(from, sizeof(int));
+        relinkJumpOrCall<BranchType_JMP, noFlush(repatch)>(reinterpret_cast<int*>(from), reinterpret_cast<const int*>(from), to);
+        if constexpr((*repatch).contains(RepatchingFlag::Flush))
```
Can we just add an `->` operator?

justinmichaud replied:
This must return a pointer. but we construct a new OptionsSet. I couldn't find a pattern for that online, but if you know of a trick, then this would be much nicer

kmiller68 replied:
Oh yeah good point. I guess you could add an `operator->() { return this; }` to OptionsSet but that's also kinda gross...

justinmichaud replied:
Oh shoot, I missed that! Done

kmiller68 reviewed (commented):
Agree with Mark's comments. Also left a comment.

justinmichaud committed:
Refactor jitMemcpy into atomic and non-atomic versions
https://bugs.webkit.org/show_bug.cgi?id=293649

Reviewed by Keith Miller.

On ARMv7, there are some concurrent repatching issues because of
unaligned writes to the jit region. Instead of repatching atomically whenever it
is easy, we should be explicit about when we expect writes to the jit
region to tear or not.

Canonical link: https://commits.webkit.org/299398@main
